balancer/rls/internal/cache/cache.go

```diff
@@ -0,0 +1,237 @@
+	// entry with this field in the past is not used to process picks.
+	ExpiryTime time.Time
+	// BackoffExpiryTime is the absolute time at which an entry which has gone
+	// through backoff stops being valid.  When an RLS request fails, this is
+	// set to the current time plus twice the backoff time. The cache expiry
+	// timer will only delete entries for which both ExpiryTime and
+	// BackoffExpiryTime are in the past.
+	BackoffExpiryTime time.Time
+	// StaleTime is the absolute time after which this entry will be
+	// proactively refreshed if we receive a request for it. When an RLS
+	// request succeeds, this is set to the current time plus the stale_age
+	// from the LB policy config.
+	StaleTime time.Time
+	// BackoffTime is the absolute time at which the backoff period for this
+	// entry ends. The backoff timer is setup with this value. No new RLS
+	// requests are sent out for this entry until the backoff period ends.
+	BackoffTime time.Time
+	// EarliestEvictTime is the absolute time before which this entry should
+	// not be evicted from the cache. This is set to a default value of 5
+	// seconds when the entry is created. This is required to make sure that a
+	// new entry added to the cache is not evicted before the RLS response
+	// arrives (usually when the cache is too small).
+	EarliestEvictTime time.Time
+	// CallStatus stores the RPC status of the previous RLS request. Picks for
+	// entries with a non-OK status are failed with the error stored here.
+	CallStatus *status.Status
+	// Backoff contains all backoff related state. When an RLS request
+	// succeeds, backoff state is reset.
+	Backoff *BackoffState
+	// HeaderData is received in an RLS response and is to be sent in the
+	// X-Google-RLS-Data header for matching RPCs.
+	HeaderData string
+	// TODO(easwars): Add support to store the ChildPolicy here. Need a
+	// balancerWrapper type to be implemented for this.
+}
+
+// BackoffState wraps all backoff related state associated with a cache entry.
+type BackoffState struct {
+	// Retries keeps track of the number of RLS failures, to be able to
+	// determine the amount of time to backoff before the next attempt.
+	Retries int
+	// Backoff is an exponential backoff implementation which returns the
+	// amount of time to backoff, given the number of retries.
+	Backoff backoff.Strategy
+	// Timer fires when the backoff period ends and incoming requests after
+	// this will trigger a new RLS request.
+	Timer *time.Timer
+	// Callback provided by the LB policy to be notified when the backoff timer
+	// expires. This will trigger a new picker to be returned to the
+	// ClientConn, to force queued up RPCs to be retried.
+	Callback func()
+}
+
+// lruEntry is the actual entry which is stored in the LRU cache. It requires
+// the key (in addition to the actual entry) as well, since the onEvicted
+// callback and expiry timer would need the key to perform their job.
```

Comment:
Another option would be to store `key` and `size` in `Entry` directly.  Did you consider this?